// stackify-node-apm/lib/common.js
'use strict'

/**
 * Responsible for common properties with
 * different implementation or values on
 * different platforms
 */
const path = require('path')
const Util = require('./util')
const Logger = require('./logger')
let prefixEnabled = false
let _debugDirectory = null
let _traceDirectory = null

function setDirectory () {
  let config = Util.loadStackifyConfig()
  let prefixEnv = !Util.isUndefined(process.env['STACKIFY_PREFIX_ENABLED']) && process.env['STACKIFY_PREFIX_ENABLED']
    ? process.env['STACKIFY_PREFIX_ENABLED'] : null

  if (prefixEnv) {
    prefixEnabled = Boolean(JSON.parse(prefixEnv))
  } else if (config && config.prefix_enabled) {
    prefixEnabled = Boolean(JSON.parse(config.prefix_enabled))
  }

  if (process.platform === 'win32') {
    if (!Util.isUndefined(process.env['ProgramData']) &&
      process.env['ProgramData'] &&
      Util.isString(process.env['ProgramData']) && (process.env['ProgramData'] !== 'null' && process.env['ProgramData'] !== '')
    ) {
      if (prefixEnabled) {
        _debugDirectory = path.join(process.env['ProgramData'], 'Stackify', 'Agent', 'debug')
        _traceDirectory = path.join(process.env['ProgramData'], 'Stackify', 'Agent', 'log')
      } else {
        _debugDirectory = path.join(process.env['ProgramData'], 'Stackify', 'NodeJs', 'log')
        _traceDirectory = path.join(process.env['ProgramData'], 'Stackify', 'NodeJs', 'log')
      }
    } else if (!Util.isUndefined(process.env['StackifyPath'])) {
      const parsedStackifyPath = path.parse(process.env['StackifyPath'])
      const rootDir = parsedStackifyPath['root']
      if (rootDir) {
        if (prefixEnabled) {
          _debugDirectory = path.join(rootDir, 'ProgramData', 'Stackify', 'Agent', 'Log')
          _traceDirectory = path.join(rootDir, 'ProgramData', 'Stackify', 'Agent', 'Log')
        } else {
          _debugDirectory = path.join(rootDir, 'ProgramData', 'Stackify', 'NodeJs', 'Log')
          _traceDirectory = path.join(rootDir, 'ProgramData', 'Stackify', 'NodeJs', 'Log')
        }
      }
    } else {
      Logger.error('Common.js - _debugDirectory / _traceDirectory - No directory found')
    }
  } else {
    // macos/linux
    if (prefixEnabled) {
      _debugDirectory = '/usr/local/prefix/debug'
      _traceDirectory = '/usr/local/prefix/log'
    } else {
      _debugDirectory = '/usr/local/stackify/stackify-node-apm/log'
      _traceDirectory = '/usr/local/stackify/stackify-node-apm/log'
    }
  }
}

setDirectory()

const Common = module.exports = {
  PREFIX_ENABLED: prefixEnabled,
  LOG_TRACES_DIR: _traceDirectory,
  LOG_DEBUG_DIR: _debugDirectory,
  NODE_VERSION: process.versions.node,
  PLATFORM: process.platform,
  LOG_FILE_BASE_NAME: 'stackify-node-apm',
  LOG_FILE_EXTENSION: '.log'
}

Common.LOG_FILE_NAME = Common.LOG_FILE_BASE_NAME + Common.LOG_FILE_EXTENSION
